pyvortex/src/array.rs

```diff
@@ -190,6 +193,69 @@ impl PyArray {
             .map(|arr| PyArray { inner: arr })
     }
 
+    /// Retrieve a row by its index.
+    ///
+    /// Parameters
+    /// ----------
+    /// index : :class:`int`
+    ///     The index of interest. Must be greater than or equal to zero and less than the length of
+    ///     this array.
+    ///
+    /// Returns
+    /// -------
+    /// :class:`Any`
+    ///
+    /// Examples
+    /// --------
+    ///
+    /// Retrieve the third element from an array of strings:
+    ///
+    /// >>> vortex.encoding.array(["hello", "goodbye", "it", "is"]).scalar_at(2)
+    /// 'it'
+    ///
+    /// Retrieve the last element from an array of integers:
+    ///
+    /// >>> vortex.encoding.array([10, 42, 999, 1992]).scalar_at(3)
+    /// 1992
+    ///
+    /// Retrieve an element from an array of structures:
+    ///
+    /// >>> array = vortex.encoding.array([
+    /// ...     {'name': 'Joseph', 'age': 25},
+    /// ...     {'name': 'Narendra', 'age': 31},
+    /// ...     {'name': 'Angela', 'age': 33},
+    /// ...     None,
+    /// ...     {'name': 'Mikhail', 'age': 57},
+    /// ... ])
+    /// >>> array.scalar_at(2)
+    /// {'age': 33, 'name': 'Angela'}
+    ///
+    /// Retrieve a missing element from an array of structures:
+    ///
+    /// >>> array.scalar_at(3) is None
+    /// True
+    ///
+    /// Out of bounds accesses are prohibited:
+    ///
+    /// >>> vortex.encoding.array([10, 42, 999, 1992]).scalar_at(10)
+    /// Traceback (most recent call last):
+    /// ...
+    /// ValueError: index 10 out of bounds from 0 to 4
+    /// ...
+    ///
+    /// Unlike Python, negative indices are not supported:
+    ///
+    /// >>> vortex.encoding.array([10, 42, 999, 1992]).scalar_at(-2)
+    /// Traceback (most recent call last):
+    /// ...
+    /// OverflowError: can't convert negative int to unsigned
+    ///
+    fn scalar_at(&self, index: &Bound<PyInt>) -> PyResult<PyObject> {
```

Comment:
how about we return a python scalar object that has `as_py()` method like arrow?

Reply:
The main argument being that for binary, utf8, and list we can return without a potentially large allocation? IMO, constructing a Scalar is expensive because of the type, though that’s admittedly dwarfed by copying a big string. I wonder if we can do a best of both worlds where the variable length stuff is returned as a view but fixed width things go straight to Python.

Reply:
I think you’re right that would be the best

Reply:
Okay, I added `scalar.rs` which has views for Buffer, BufferString, List, and Struct. I did not flesh these out at all, they just have a method `into_python` which copies into Python.

I think we maybe want the buffer and buffer string to have a way to produce a `memoryview` which would permit, for example, regex matching.

For now though, we have a backwards compatible interface for users and we can focus on the rest of the UX punchlist. Does that seem reasonable to you? I added a todo to the bottom of the UX punchlist https://github.com/spiraldb/vortex/issues/1072